### PR TITLE
[infra-proxy-service] Roles list API bug fixes

### PR DESCRIPTION
**Pull request by vsingh-msys**
### :nut_and_bolt: Description: What code changed, and why?

- Fixes the role list API by counting the length of rows instead of accessing the total from resultset.
- Remove the Role struct from roles.go and considered the role struct after `EnvRunList` got added. [`ac86f72` (#180)](https://github.com/go-chef/chef/pull/180/commits/ac86f72cf00f081ea5db6ae494278dea4a4a4bce)

### :chains: Related Resources
Fixes: #4619

### :+1: Definition of Done

- Fixes role list API failing for some chef server data.

### :athletic_shoe: How to Build and Test the Change
- `rebuild components/infra-proxy-service`
- `rebuild components/automate-gateway`

Steps to test:

- To add data using 3rd internal steps https://github.com/chef/automate/blob/master/dev-docs/adding-data/adding_test_data.md#adding-data-to-infra-views

- Roles list API:

```
curl -sSkH "api-token: $(get_admin_token)" https://a2-dev.test/api/v0/infra/servers/chef-server-dev-test/orgs/chef-org-dev/roles?pretty
{
  "roles": [
    {
      "name": "chef-client-update",
      "description": "",
      "environments": [
        "_default"
      ]
    },
    {
      "name": "web-server",
      "description": "",
      "environments": [
        "_default"
      ]
    }
  ]
}
```

### :white_check_mark: Checklist

- [x] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [ ] Tests added/updated?
- [ ] Docs added/updated?
- [x] All commits have been signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### :camera: Screenshots, if applicable
Signed-off-by: Vivek Singh <vivek.singh@msystechnologies.com>

### CODE REVIEW

File: components/infra-proxy-service/server/roles.go [hunk 1]
```diff
@@ -235,7 +223,7 @@ func (s *Server) UpdateRole(ctx context.Context, req *request.UpdateRole) (*resp
 
 // fromAPIToListRoles a response.Roles from a struct of RoleList
 func fromAPIToListRoles(result RoleListResult) []*response.RoleListItem {
-	cl := make([]*response.RoleListItem, result.Total)
+	cl := make([]*response.RoleListItem, len(result.Rows))
```
Comment (vsingh-msys) [inline components/infra-proxy-service/server/roles.go:226]: Observe inconsistency in chef server data:
```
knife raw /search/role
{
  "total": 4,
  "start": 0,
  "rows": [

  ]
}
```
To fix this calculating the rows length.

File: components/infra-proxy-service/server/roles.go [hunk 2]
```diff
@@ -20,23 +20,11 @@ type RunList []string
 // EnvRunList represents the recipes and roles with environment specified for a node or as part of a role.
```
Comment (kalroy) [inline components/infra-proxy-service/server/roles.go:20]: RunList and EnvRunList are defined under chef package. We may not need to redefine it here and refer the type defined in chef package.

**Review by kalroy — CHANGES_REQUESTED** [review]
Added some feedbacks to look into.

File: components/infra-proxy-service/server/roles.go [hunk 3]
```diff
@@ -20,23 +20,11 @@ type RunList []string
...
 type EnvRunList map[string]RunList
```
[the remark this answers is not in the document]
Comment (vsingh-msys) [reply components/infra-proxy-service/server/roles.go:21]: yeah, thanks!

**Review by vivekshankar1 — APPROVED** [review]

**Review by kalroy — APPROVED** [review]
Good job. Thank you!! 👏

**Review by msorens — CHANGES_REQUESTED** [review]
Was going to suggest there should be a unit test to prove this change, but roles.go does not have any unit tests at all; please add.

**Comment by vsingh-msys** [comment]
This issue due to chef infra server corrupted data, fixes by `chef-server-ctl reindex`, Also list will be changing by search API in PR https://github.com/chef/automate/pull/4658

**Pull request closed without merging**